File: scripts/lib/normalize.py
"""문자열 정규화 — 앱(JS)과 배치(Python)가 동일하게 구현해야 하는 규칙.

taxonomy.yaml `normalization.steps`가 단일 소스이며,
scripts/normalize_test.py의 구현과 문자 단위로 같은 결과를 낸다.

배치가 이 모듈을 쓰는 이유는 blocklist 매칭이다.
"죽 고 싶", "죽고싶ㅠㅠㅠ", "죽!고!싶!" 은 모두 같은 문자열로 접혀서
띄어쓰기·반복·문장부호를 이용한 우회를 막는다.
"""

from __future__ import annotations

import re
from collections.abc import Iterable, Sequence

# 공백·문장부호·이모지 제거. 한글 자모(ㄱ-ㅎ ㅏ-ㅣ)는 감정 표현이라 보존한다.
_NON_WORD = re.compile(r"[^\w가-힣ㄱ-ㅎㅏ-ㅣa-z0-9]")

# 3회 이상 반복되는 문자를 2회로 축약 ("짜증나아아아" → "짜증나아")
_REPEATS = re.compile(r"(.)\1{2,}")


def normalize(text: str) -> str:
    """비교용 정규화 문자열을 반환한다."""
    if not text:
        return ""
    t = text.lower()
    t = _NON_WORD.sub("", t)
    t = _REPEATS.sub(r"\1\1", t)
    return t.replace("ㅜ", "ㅠ")


def matched_terms(haystack: str, terms: Iterable[str]) -> list[str]:
    """haystack(원문)에 포함된 terms를 정규화 기준으로 찾아 반환한다.

    걸러낸 이유를 로그·리포트에 남길 수 있도록 불리언이 아니라 목록을 준다.
    """
    normalized = normalize(haystack)
    if not normalized:
        return []
    hits: list[str] = []
    for term in terms:
        needle = normalize(term)
        if needle and needle in normalized:
            hits.append(term)
    return hits


def contains_any(haystack: str, terms: Sequence[str]) -> bool:
    """terms 중 하나라도 포함되면 True (단락 평가로 조기 종료)."""
    normalized = normalize(haystack)
    if not normalized:
        return False
    return any((needle := normalize(t)) and needle in normalized for t in terms)

File: app/src/lib/normalize.js
/**
 * 입력 정규화 — scripts/lib/normalize.py와 문자 단위로 같은 결과를 내야 한다.
 *
 * 왜 같아야 하나:
 *   배치는 이 규칙으로 blocklist를 적용하고, 앱은 같은 규칙으로 위기 키워드와
 *   감정 키워드를 매칭한다. 두 구현이 어긋나면 배치가 거른 표현을 앱이 통과시키거나
 *   그 반대가 된다. taxonomy.yaml normalization.steps가 단일 소스다.
 *
 * 프로토타입 코드를 그대로 쓰지 않은 이유:
 *   emotion-prototype.jsx는 /[^\w가-힣ㄱ-ㅎㅏ-ㅣa-z0-9]/g 를 썼는데,
 *   JS의 \w는 [A-Za-z0-9_]뿐이라 Python의 \w(유니코드 전체 문자)보다 훨씬 좁다.
 *   한글만 있는 입력에서는 결과가 같지만 한자·가나·악센트 문자가 섞이면 갈린다.
 *   Python의 \w에 대응시키려면 유니코드 속성 이스케이프가 필요하다.
 *
 *   Python  re: \w == 유니코드 단어 문자 + '_'
 *   JS     대응: \p{L}(문자) + \p{N}(숫자) + '_'  ※ u 플래그 필수
 *
 *   패턴에 남아 있던 가-힣·ㄱ-ㅎ·ㅏ-ㅣ·a-z0-9는 \p{L}\p{N}에 이미 포함되므로 뺐다.
 *   (한글 자모 ㄱ-ㅎ, ㅏ-ㅣ는 유니코드 카테고리 Lo라 \p{L}에 들어온다)
 */

const NON_WORD = /[^\p{L}\p{N}_]/gu;
const REPEATS = /(.)\1{2,}/gu;

/** 비교용 정규화 문자열을 반환한다. */
export function normalize(text) {
  if (!text) return "";
  return text
    .toLowerCase()
    .replace(NON_WORD, "")
    .replace(REPEATS, "$1$1")
    .replace(/ㅜ/gu, "ㅠ");
}

/**
 * haystack(원문)에 포함된 terms를 정규화 기준으로 찾아 반환한다.
 * 걸러낸 이유를 남길 수 있도록 불리언이 아니라 목록을 준다.
 */
export function matchedTerms(haystack, terms) {
  const normalized = normalize(haystack);
  if (!normalized) return [];
  const hits = [];
  for (const term of terms) {
    const needle = normalize(term);
    if (needle && normalized.includes(needle)) hits.push(term);
  }
  return hits;
}

/** terms 중 하나라도 포함되면 true (조기 종료). */
export function containsAny(haystack, terms) {
  const normalized = normalize(haystack);
  if (!normalized) return false;
  return terms.some((t) => {
    const needle = normalize(t);
    return needle && normalized.includes(needle);
  });
}
